File: android/app/src/main/java/com/easyweb/app/engine/Render.kt
// b站：绝望彻彻底底的绝望 [android/app/src/main/java/com/easyweb/app/engine/Render.kt]

package com.easyweb.app.engine

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import kotlin.math.roundToInt

// 渲染器
class Renderer(
    private val canvas: Canvas,
    private val scrollX: Float = 0.0f,
    private val scrollY: Float = 0.0f,
    private val viewportWidth: Float,
    private val viewportHeight: Float,
    private val scaleFactor: Float = 1.0f
) {
    private val bgPaint = Paint()
    private val borderPaint = Paint()
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    // 渲染整个布局树
    fun render(layoutRoot: LayoutBox) {
        // 清空画布为白色背景
        canvas.drawColor(Color.WHITE)
        // 应用缩放
        canvas.save()
        canvas.scale(scaleFactor, scaleFactor)
        // 递归渲染盒子
        renderBox(layoutRoot)
        // 恢复画布
        canvas.restore()
    }

    // 渲染单个盒子
    private fun renderBox(box: LayoutBox) {
        // 跳过屏幕外的盒子
        val screenRect = Rect(-scrollX, -scrollY, viewportWidth, viewportHeight)
        if (!rectsOverlap(box.rect, screenRect)) return

        // 渲染背景和边框
        renderBackground(box)
        renderBorder(box)

        // 渲染文本
        if (box.boxType == BoxType.Text) {
            renderText(box)
        }

        // 渲染子元素
        for (child in box.children) {
            renderBox(child)
        }
    }

    // 渲染背景
    private fun renderBackground(box: LayoutBox) {
        val style = box.style ?: return
        val bg = style.color("background-color", 0L)
        val alpha = alphaOf(bg)
        if (alpha > 0) {
            val pa = paddingArea(box)
            val x = (pa.x - scrollX).roundToInt()
            val y = (pa.y - scrollY).roundToInt()
            val w = pa.width.roundToInt()
            val h = pa.height.roundToInt()
            if (w > 0 && h > 0) {
                bgPaint.color = Color.argb(
                    alpha.coerceIn(0, 255),
                    redOf(bg).coerceIn(0, 255),
                    greenOf(bg).coerceIn(0, 255),
                    blueOf(bg).coerceIn(0, 255)
                )
                bgPaint.style = Paint.Style.FILL
                canvas.drawRect(x.toFloat(), y.toFloat(), (x + w).toFloat(), (y + h).toFloat(), bgPaint)
            }
        }
    }

    // 渲染边框
    private fun renderBorder(box: LayoutBox) {
        if (box.border.top == 0.0f && box.border.right == 0.0f
            && box.border.bottom == 0.0f && box.border.left == 0.0f) return

        val borderColor = box.style?.color("border-color", toArgb(0, 0, 0, 255)) ?: toArgb(0, 0, 0, 255)
        borderPaint.color = Color.argb(
            alphaOf(borderColor).coerceIn(0, 255),
            redOf(borderColor).coerceIn(0, 255),
            greenOf(borderColor).coerceIn(0, 255),
            blueOf(borderColor).coerceIn(0, 255)
        )
        borderPaint.style = Paint.Style.FILL

        val r = box.rect
        val x = (r.x - scrollX).toFloat()
        val y = (r.y - scrollY).toFloat()
        val w = r.width
        val h = r.height
        val bTop = box.border.top
        val bRight = box.border.right
        val bBottom = box.border.bottom
        val bLeft = box.border.left

        // 上边框
        if (bTop > 0.0f) {
            canvas.drawRect(x, y, x + w, y + bTop, borderPaint)
        }
        // 下边框
        if (bBottom > 0.0f) {
            canvas.drawRect(x, y + h - bBottom, x + w, y + h, borderPaint)
        }
        // 左边框
        if (bLeft > 0.0f) {
            canvas.drawRect(x, y, x + bLeft, y + h, borderPaint)
        }
        // 右边框
        if (bRight > 0.0f) {
            canvas.drawRect(x + w - bRight, y, x + w, y + h, borderPaint)
        }
    }

    // 渲染文本
    private fun renderText(box: LayoutBox) {
        val node = box.domNode ?: return
        if (node.kind != NodeKind.Text) return
        val text = node.text?.trim() ?: return
        if (text.isEmpty()) return

        val color = box.style?.color("color", toArgb(0, 0, 0, 255)) ?: toArgb(0, 0, 0, 255)
        val fontSize = box.style?.length("font-size", 16.0f) ?: 16.0f

        textPaint.color = Color.argb(
            alphaOf(color).coerceIn(0, 255),
            redOf(color).coerceIn(0, 255),
            greenOf(color).coerceIn(0, 255),
            blueOf(color).coerceIn(0, 255)
        )
        textPaint.textSize = fontSize
        textPaint.typeface = Typeface.DEFAULT

        val x = box.rect.x - scrollX
        val y = box.rect.y - scrollY

        // 使用 Android Canvas 绘制文本
        canvas.drawText(text, x, y + fontSize, textPaint)
    }
}

// 判断两个矩形是否重叠
private fun rectsOverlap(a: Rect, b: Rect): Boolean {
    return a.x < b.x + b.width && a.x + a.width > b.x
        && a.y < b.y + b.height && a.y + a.height > b.y
}

File: src/render.rs
use sdl2::pixels::Color;
use sdl2::rect::Rect as SdlRect;
use sdl2::render::Canvas;
use sdl2::ttf::Font;
use sdl2::video::Window;

use crate::dom::NodeKind;
use crate::layout::{BoxType, LayoutBox, Rect};

/// Render the layout tree to an SDL2 canvas.
pub fn render(
    canvas: &mut Canvas<Window>,
    font: &Font,
    layout_root: &LayoutBox,
    scroll_x: f32,
    scroll_y: f32,
    viewport_w: f32,
    viewport_h: f32,
) {
    canvas.set_draw_color(Color::RGB(255, 255, 255));
    canvas.clear();
    render_box(canvas, font, layout_root, scroll_x, scroll_y, viewport_w, viewport_h);
    canvas.present();
}

fn render_box(
    canvas: &mut Canvas<Window>,
    font: &Font,
    box_: &LayoutBox,
    scroll_x: f32,
    scroll_y: f32,
    viewport_w: f32,
    viewport_h: f32,
) {
    // Skip if off-screen
    let screen_rect = Rect::new(-scroll_x, -scroll_y, viewport_w, viewport_h);
    if !rects_overlap(&box_.rect, &screen_rect) {
        return;
    }

    // Render background and border
    render_background(canvas, box_, scroll_x, scroll_y);
    render_border(canvas, box_, scroll_x, scroll_y);

    // Render text content
    if box_.box_type == BoxType::Text {
        render_text(canvas, font, box_, scroll_x, scroll_y);
    }

    // Render children
    for child in &box_.children {
        render_box(canvas, font, child, scroll_x, scroll_y, viewport_w, viewport_h);
    }
}

fn render_background(canvas: &mut Canvas<Window>, box_: &LayoutBox, scroll_x: f32, scroll_y: f32) {
    if let Some(ref style) = box_.style {
        let bg = style.color("background-color", (0, 0, 0, 0));
        if bg.3 > 0 {
            let pa = box_.padding_area();
            let x = (pa.x - scroll_x).round() as i32;
            let y = (pa.y - scroll_y).round() as i32;
            let w = pa.width.round() as u32;
            let h = pa.height.round() as u32;

            if w > 0 && h > 0 {
                canvas.set_draw_color(Color::RGBA(bg.0, bg.1, bg.2, bg.3));
                let _ = canvas.fill_rect(SdlRect::new(x, y, w, h));
            }
        }
    }
}

fn render_border(canvas: &mut Canvas<Window>, box_: &LayoutBox, scroll_x: f32, scroll_y: f32) {
    if box_.border.top == 0.0 && box_.border.right == 0.0
        && box_.border.bottom == 0.0 && box_.border.left == 0.0
    {
        return;
    }

    let border_color = box_.style
        .as_ref()
        .map(|s| s.color("border-color", (0, 0, 0, 255)))
        .unwrap_or((0, 0, 0, 255));

    canvas.set_draw_color(Color::RGBA(border_color.0, border_color.1, border_color.2, border_color.3));

    let r = &box_.rect;
    let x = (r.x - scroll_x).round() as i32;
    let y = (r.y - scroll_y).round() as i32;
    let w = r.width.round() as i32;
    let h = r.height.round() as i32;
    let b_top = box_.border.top.round() as u32;
    let b_right = box_.border.right.round() as u32;
    let b_bottom = box_.border.bottom.round() as u32;
    let b_left = box_.border.left.round() as u32;

    // Top
    if b_top > 0 {
        let _ = canvas.fill_rect(SdlRect::new(x, y, w as u32, b_top));
    }
    // Bottom
    if b_bottom > 0 {
        let _ = canvas.fill_rect(SdlRect::new(x, y + h - b_bottom as i32, w as u32, b_bottom));
    }
    // Left
    if b_left > 0 {
        let _ = canvas.fill_rect(SdlRect::new(x, y, b_left, h as u32));
    }
    // Right
    if b_right > 0 {
        let _ = canvas.fill_rect(SdlRect::new(x + w - b_right as i32, y, b_right, h as u32));
    }
}

fn render_text(canvas: &mut Canvas<Window>, font: &Font, box_: &LayoutBox, scroll_x: f32, scroll_y: f32) {
    if let Some(ref node) = box_.dom_node {
        if let NodeKind::Text(t) = &node.borrow().kind {
            let text = t.trim();
            if text.is_empty() {
                return;
            }

            // Check if we have a cached texture
            let color = box_.style
                .as_ref()
                .map(|s| s.color("color", (0, 0, 0, 255)))
                .unwrap_or((0, 0, 0, 255));

            let x = (box_.rect.x - scroll_x).round() as i32;
            let y = (box_.rect.y - scroll_y).round() as i32;

            // Render text using SDL2_ttf blit
            if let Ok(surface) = font
                .render(text)
                .blended(Color::RGBA(color.0, color.1, color.2, color.3))
            {
                let w = surface.width();
                let h = surface.height();

                let texture_creator = canvas.texture_creator();
                let texture_result = surface.as_texture(&texture_creator);
                if let Ok(texture) = texture_result {
                    let _ = canvas.copy(
                        &texture,
                        None,
                        SdlRect::new(x, y, w, h),
                    );
                }
            }
        }
    }
}

fn rects_overlap(a: &Rect, b: &Rect) -> bool {
    a.x < b.x + b.width && a.x + a.width > b.x && a.y < b.y + b.height && a.y + a.height > b.y
}

/// Calculate the total height of the layout tree for scrolling.
pub fn layout_height(root: &LayoutBox) -> f32 {
    root.rect.y + root.rect.height
}